### PR TITLE
Ignore Numericising Specific Columns in get_all_records

### LINKED ISSUE #581: specify text columns so get_all_records does not run numericise
I have an id column in my spreadsheet that contains alphanumeric strings. Some of these strings are being interpreted as numbers. Take the following example:
```
from gspread.utils import numericise_all

id_column = ['8085807dc0000000', '8085807e40000000', '8085807ec0000000', '8085805e40000000']
numericise_all(id_column)
```
Returns:
```
['8085807dc0000000', inf, '8085807ec0000000', inf]
```
Of course this means that the id column is no longer useful for uniquely identifying records.

I would like to be able to specify that `id_column` should be kept as a string. Maybe something like `sheet.get_all_records(text_columns=['id_column'])`?

### PR DESCRIPTION
**Pull request by benjamindhimes**
Allows passing in a list of column indices (1-indexed) to avoid numericising them in the get_all_records method, also allows for the special case ['all'] to avoid numericising all columns. Fixes #581,  #568, #231 Improvement over my previously-closed PR of #598.

### CODE REVIEW

**Comment by vijay-shanker** [comment]
@burnash this looks like an interesting problem, but the PR has been closed, should be picked up, can i ?

**Comment by bhimesuber** [comment]
@vijay-shanker this PR is not closed

**Comment by burnash** [comment]
@benjamindhimes Thank you very much for your contribution. I'm really sorry for the late response and I appreciate your persistence. I'll review the PR shortly.

File: gspread/models.py [hunk 1]
```diff
@@ -744,14 +744,15 @@ def get_all_records(
         head=1,
         default_blank="",
         allow_underscores_in_numeric_literals=False,
+        numericise_ignore=[]
```
Comment (burnash) [inline gspread/models.py:747]: A [mutable default argument](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) may lead to surprising behavior.

Comment (benjamindhimes) [reply gspread/models.py]: You're quite right. Fixed.

File: gspread/utils.py [hunk 2]
```diff
@@ -141,10 +141,27 @@ def numericise(value, empty2zero=False, default_blank="", allow_underscores_in_n
 
     return value
 
+def numericise_all(input, empty2zero=False, default_blank="", allow_underscores_in_numeric_literals=False,
+                   ignore=[]):
```
Comment (burnash) [inline gspread/utils.py:145]: Sorry, the same problem is here.

File: gspread/utils.py [hunk 3]
```diff
@@ -141,10 +141,27 @@ def numericise(value, empty2zero=False, default_blank="", allow_underscores_in_n
...
+    :param input: list row
+    :param empty2zero: bool indicating whether or not to return empty cells as 0, default False
+    :param default_blank: str determining which value to use for blank cells, default empty str
+    :param allow_underscores_in_numeric_literals: bool indicating whether or not to allow visual underscores
+                                                  in numeric literals
+    :param ignore: list of ints of indices of the row (index 1) to ignore numericising
+    """
+    _numericise = partial(numericise, empty2zero=empty2zero, default_blank=default_blank,
+                          allow_underscores_in_numeric_literals=allow_underscores_in_numeric_literals)
+    if not ignore:
```
Comment (aiguofer) [inline gspread/utils.py:158]: You could just set `ignore = ignore or []` and get rid of the `if` statement

File: gspread/models.py [hunk 4]
```diff
@@ -778,15 +783,21 @@ def get_all_records(
...
-                row,
-                empty2zero,
-                default_blank,
-                allow_underscores_in_numeric_literals,
-            )
-            for row in data[idx + 1:]
-        ]
+
+        if numericise_ignore == ['all']:
+            values = [row for row in data[idx + 1]]
```
Comment (burnash) [inline gspread/models.py:788]: It looks like in this case it only iterates over one row. Shouldn't it be `data[idx + 1:]` like in an `else` clause?

**Comment by burnash** [comment]
Works well. Thank you very much! Kudos for expanding the docstring for `numericise_all()`

**Comment by mmsbr** [comment]
Hello friends! How can I use this fix in my code? I'm using gspread for python3 and I really need the leading zeros

**Comment by mmsbr** [comment]
Just to answer my own question and to help others that may have the same problem. Now you can keep leading zeros just using get_all_records() like this: get_all_records(numericise_ignore=['all']) and gspread won't convert your strings! Thank so much for this @benjamindhimes !!! You are awesome!

**Comment by Ehayanarle** [comment]
@mmsbr I was looking for the solving of a problem for a couple of days but it was so easy. I should read the fu***** documentation and code of the module. Thank you very much for the solution!

**Pull request merged**